Introduce new parallel::fullydistributed::Triangulation

peterrum opened the pull request:
This PR is part of the effort to introduce the new `parallel::fullydistributed::Triangulation` (see #8558).

In this PR, the actual `parallel::fullydistributed::Triangulation` is added. Furthermore some tests are added and a single basic function is added to create the `parallel::fullydistributed::ConstructionData`.

Ping: @tjhei @bangerth @kronbichler @masterleinad

peterrum commented:
Note: I am still refactoring and cleaning up the code. However, what you can see is that the majority of this PR are headers and tests. The relevant files are:
- include/deal.II/distributed/fullydistributed_tria_util.h (446 loc)
- source/distributed/fullydistributed_tria.cc (562 loc)

I'll signal when the code is ready for review!

bangerth commented on include/deal.II/distributed/fullydistributed_tria.h:
```diff
@@ -0,0 +1,237 @@
+// ---------------------------------------------------------------------
+
+#ifndef dealii_fullydistributed_tria_h
+#define dealii_fullydistributed_tria_h
```
We generally use `CamelCase` for class names and `under_scores` for variable and file names. Can you rename the file to `full_distributed_tria.h` and adjust the `#include` guard here?

peterrum replied:
I have done it locally. But did you remove the y intentionally?

bangerth commented on include/deal.II/distributed/fullydistributed_tria.h:
```diff
@@ -0,0 +1,237 @@
+#  include <mpi.h>
+#endif
+
+#ifdef DEAL_II_WITH_P4EST
```
Do you need these headers? I think you can skip everything p4est related in this file.

peterrum replied:
These were a copy of the includes in `tria.h`. I have removed them.

bangerth commented on include/deal.II/distributed/fullydistributed_tria.h:
```diff
@@ -0,0 +1,237 @@
+    namespace Policy
+    {
+      template <typename>
+      class ParallelDistributed;
```
There's been a recent patch to enclose all forward declarations in `#ifndef DOXYGEN` -- I don't have the number at hand, but it should be easy to find.

peterrum replied:
Done!

bangerth commented on include/deal.II/distributed/fullydistributed_tria.h:
```diff
@@ -0,0 +1,237 @@
+namespace GridTools
+{
+  template <typename CellIterator>
+  struct PeriodicFacePair;
```
All the way down to here.

peterrum replied:
Done!

bangerth commented on include/deal.II/distributed/fullydistributed_tria.h:
```diff
@@ -0,0 +1,237 @@
+  {
+    struct CellInfo
+    {
+      CellInfo() = default;
```
documentation missing for the class and its members

peterrum replied:
I know: WIP.

tjhei commented on include/deal.II/distributed/fullydistributed_tria.h:
```diff
@@ -0,0 +1,237 @@
+      };
+
+
+      Triangulation(MPI_Comm mpi_communicator,
```
please mark this constructor explicit

peterrum replied:
Done. I guess, I can also create a PR to make the constructor of `p:d:t` explicit?

masterleinad replied:
Yes, that's a good idea.

tjhei commented on include/deal.II/distributed/fullydistributed_tria.h:
```diff
@@ -0,0 +1,237 @@
+
+
+      Triangulation(MPI_Comm mpi_communicator,
+                    Settings settings_ = default_setting);
```
we typically don't use underscores at the end of variable names.

peterrum replied:
I have removed it. See also PR #8627

tjhei commented on include/deal.II/distributed/fullydistributed_tria.h:
```diff
@@ -0,0 +1,237 @@
+
+
+    template <int dim, int spacedim>
+    struct ConstructionData
```
all classes, namespaces, variables, and functions will need doxygen documentation (I know this is still WIP, but still. :-) )

peterrum replied:
I know: WIP! ;)

tjhei commented on include/deal.II/distributed/fullydistributed_tria.h:
```diff
@@ -0,0 +1,237 @@
+        typename dealii::Triangulation<dim, spacedim>::CellStatus CellStatus;
+
+      void
+      reinit(ConstructionData<dim, spacedim> &construction_data);
```
make the argument const (here and in many other places)

peterrum replied:
Fixed here. More to come: WIP ;)

tjhei commented on source/distributed/fullydistributed_tria.cc:
```diff
@@ -0,0 +1,562 @@
+      if (construction_data.cells.size() == 0) // no
+        {
+          // 1) create a dummy hypercube
+          Point<spacedim> p1, p2;
```
why do you need this? otherwise the triangulation will return ``empty()==true`` which breaks other code?

peterrum replied:
The library will break if I remove it because it implicitly assumes that there is at least one cell on each process (which is naturally the case for `p:d:t`, since the coarse grid is duplicated; even if number of procs is bigger than number of active cells).

bangerth replied:
If a user provides an invalid input argument (such as an empty `construction_data`), then catch that via an assertion, rather than working around it.

bangerth replied:
Or are you saying that it's a valid thing to do because one might not have a locally owned cell on all processors? I'd much rather us fix the breakages that result, rather than work around things by creating "artificial" cells that don't actually match anything elsewhere -- you're going to get into all sorts of problems downstream with your approach because you now have these kinds of cells. For example, you can now no longer rely on the fact that artificial cells are real cells with real coordinates at some level of the hierarchy. In other words, you can't trust anything any more about the geometry of the mesh.

peterrum replied:
That a program aborts if one process does not have any cells is in my opinion too restrictive: given adaptivity, you might start with a very coarse mesh and then refine repeatedly (in this case the program would crash at the beginning).

I agree with you fear regarding the *artificial* information about the geometry. The motivation for doing the way I did it was the assumption that you would/should not access artificial cells.

I will have look into an alternative approach and I will write you. My fear is that both `create_triangulation` and `update_cache` (of `dealii::Triangulation`) will have issues with a mesh definition without cells. Would it be a solution to enable `dealii::Triangulation` to have zero cells? If yes, I would open a new PR and write a test.

peterrum replied:
@bangerth How do we want to proceed with this issue? Do we want in the case that there are processes without locally relevant cells that these processes have a single artificial cell (as it is implemented right now) or that they have no cells.

Both have pros and cons. In the case of no cells, (at least) a couple of asserts have to disabled e.g. in `DoFHandlerPolicy` so that they can handle triangulations without levels. I must admit I do not have a good feeling about removing those asserts. Furthermore there are - as far I as I see and first tests show - parts in the library which assume that there is at least one coarse cell...

peterrum replied:
@bangerth Would it be a solution to have a single artificial cell. But instead of creating it with the help of `GridGenerator::hypercube`, I use any global coarse-grid cell and set it artificial.

tjhei commented on source/distributed/fullydistributed_tria.cc:
```diff
@@ -0,0 +1,562 @@
+        {
+          // 1) create a dummy hypercube
+          Point<spacedim> p1, p2;
+          for (unsigned int i = 0; i < dim; ++i)
```
can you replace the whole code here by ``GridGenerator::hyper_cube(*this)``?

peterrum replied:
I have done it with the help forward declaration (I really don't like forward declaration - it breaks clear dependencies).

masterleinad replied:
> I really don't like forward declaration - it breaks clear dependencies

What do you mean? I normally use forward declarations to be able to deal with dependencies properly. :slightly_smiling_face:

tjhei commented on source/distributed/fullydistributed_tria.cc:
```diff
@@ -0,0 +1,562 @@
+
+
+          // create inverse map
+          std::map<int, int> coarse_gid_to_lid;
```
wrong types here!

peterrum replied:
I know. This is older than `types::coarse_cell_id` and our discussions. I have started to update the data types yesterday and have not finished it yet. WIP ;)

tjhei commented on source/distributed/fullydistributed_tria.cc:
```diff
@@ -0,0 +1,562 @@
+            vertices, cells, subcelldata);
+
+          for (auto c = this->begin(); c != this->end(); c++)
+            c->set_all_manifold_ids(c->material_id());
```
why do you do this? Can you make sure this is documented in the class?

peterrum replied:
Ups. What I meant (I guess my test passed because `material_id` and `manifold_id` happened to be the same):
```cpp
          for (auto c = this->begin(); c != this->end(); c++)
            c->set_all_manifold_ids(c->manifold_id());
```

I need theses two lines to be able to create a `hyper_shell()` with vertices put at the right position while refining (see also dealii step-1). Setting `CellData::manifold_id` was not enough...

tjhei replied:
The user is free to use different manifolds ids on cells, faces, and edges (and this is often useful to do!). The call to ``set_all_manifold_ids`` will set this for the cell and all child objects. I assume your ConstructionData does not carry information about face/edge manifold ids? I think it should...

peterrum replied:
OK! I will extend `ConstructionData` with this information.

peterrum replied:
@tjhei I have extended `ConstructionData` with `manifold_quad_ids` and `manifold_line_ids` (and have added some comments). 

Could you please have a look if this is fine this way?

tjhei commented on source/distributed/fullydistributed_tria.cc:
```diff
@@ -0,0 +1,562 @@
+
+              if (settings & construct_multigrid_hierarchy)
+                cell->set_level_subdomain_id(
+                  dealii::numbers::artificial_subdomain_id);
```
what level_subdomain_id do we want if ``construct_multigrid_hierarchy`` is not set? Maybe also artificial?

peterrum replied:
Per default, I set all the `level_subdomain_id` of all cells to `dealii::numbers::artificial_subdomain_id`.

tjhei replied:
My point here is: you should remove the if and do this call unconditionally, so that the ids are artificial if the user is not constructing the multigrid hierarchy.

peterrum replied:
I have already removed it yesterday.

tjhei commented on source/distributed/fullydistributed_tria.cc:
```diff
@@ -0,0 +1,562 @@
+              typename dealii::Triangulation<dim, spacedim>::MeshSmoothing>(
+              dealii::Triangulation<dim>::none |
+              Triangulation<dim,
+                            spacedim>::limit_level_difference_at_vertices) :
```
Is this necessary to create the correct mesh? I wonder if it works without...

peterrum replied:
At the moment it is necessary since this is the way how I construct the `ConstructionData`. Is there a need to weaken this? Once we have hanging nodes and multigrid, we need it?

tjhei replied:
One needs to satisfy this for GMG when reconstructing the local mesh from the oracle to have ghost level neighbors set correctly. My assumption is that you construct the p:fd:Tria from an object that already contains all cells and neighbors that you will need. Hence, I believe this flag is not needed for you here.

peterrum replied:
As I said above, I have written the `ConstructionData` in such a way that it fulfills the condition`limit_level_difference_at_vertices` (by adding some artificial cells). I guess I can modify the definition of `ConstructionData` in such a way that we do not have to set this flag.

tjhei commented on include/deal.II/distributed/fullydistributed_tria.h:
```diff
@@ -0,0 +1,237 @@
+      std::vector<std::pair<types::coarse_cell_id, unsigned int>>
+        coarse_gid_to_lid;
+      std::vector<std::pair<unsigned int, types::coarse_cell_id>>
+        coarse_lid_to_gid;
```
I would have expected that it is enough to store a std::vector< types::coarse_cell_id> of length n_cells(0) here. Can this have gaps?

peterrum replied:
You are right having a simple `std::vector< types::coarse_cell_id>` is better. I have changed it.

tjhei commented on include/deal.II/distributed/fullydistributed_tria.h:
```diff
@@ -0,0 +1,237 @@
+      std::vector<std::pair<unsigned int, types::coarse_cell_id>>
+        coarse_lid_to_gid;
+
+      MPI_Comm mpi_communicator_coarse;
```
what does this communicator do?

peterrum replied:
I take this out for this PR. The plan is to re-partition the triangulation on the coarsest grid with a new MPI communicator and only to use this communicator on the coarsest level.

peterrum commented:
@tjhei @bangerth All your comments are much appreciated!

peterrum commented:
@tjhei I think `ConstructionData` and `CellInfo` now fulfill your description. 

@bangerth @tjhei  Except of the special treatment of processes without cells (which will be implemented as I agreed above with @bangerth ), I think the files `full_distributed_tria.h` and `full_distributed_tria.cc` are ready to be reviewed. I am still working on `full_distributed_tria_util.h`, however, since changes form `full_distributed_tria.h`/`full_distributed_tria.cc` directly involve changes also in `full_distributed_tria_util.h`, I would be thankful for a review of those two files, before I proceed with a premature refactoring of `full_distributed_tria_util.h`.

I will have a look at the `MeshSmoothing`-flags during refactoring of `full_distributed_tria_util.h` (since small changes during setup of `ConstructionData` have to be performed) or will postpone it to a next PR?

tjhei commented on include/deal.II/distributed/full_distributed_tria.h:
```diff
@@ -0,0 +1,315 @@
+      reinit(const ConstructionData<dim, spacedim> &construction_data);
+
+      /**
+       * @note Use the function reinit() instead of this function to create
```
"This function is not implemented for this class. Instead, use the function reinit() to create the triangulation."

tjhei commented on include/deal.II/distributed/full_distributed_tria.h:
```diff
@@ -0,0 +1,315 @@
+      compute_vertices_with_ghost_neighbors() const override;
+
+      virtual bool
+      is_multilevel_hierarchy_constructed() const override;
```
document

tjhei commented on include/deal.II/distributed/full_distributed_tria.h:
```diff
@@ -0,0 +1,315 @@
+      is_multilevel_hierarchy_constructed() const override;
+
+      virtual unsigned int
+      coarse_cell_id_to_coarse_cell_index(
```
copy documentation (or @copydoc)

tjhei commented on include/deal.II/distributed/full_distributed_tria.h:
```diff
@@ -0,0 +1,315 @@
+      coarse_cell_id_to_coarse_cell_index(
+        const types::coarse_cell_id coarse_cell_id) const override;
+
+      virtual types::coarse_cell_id
```
same here

tjhei commented on include/deal.II/distributed/full_distributed_tria.h:
```diff
@@ -0,0 +1,315 @@
+      const Settings settings;
+
+      /**
+       * Sorted list coarse-cell ids (and their index).
```
```suggestion
       * Sorted list of coarse-cell ids and their index.
```

tjhei commented on include/deal.II/distributed/full_distributed_tria.h:
```diff
@@ -0,0 +1,315 @@
+        coarse_cell_id_to_coarse_cell_index_vector;
+
+      /**
+       * List of coarse-cell ids sorted according to their index.
```
```suggestion
       * coarse-cell id for each coarse cell (stored at cell->index()).
```

tjhei commented on source/distributed/full_distributed_tria.cc:
```diff
@@ -0,0 +1,499 @@
+    namespace internal
+    {
+      /**
+       * Check if `parent` is parent of `child`.
```
```suggestion
       * Check if `parent` is the parent of `child`.
```

tjhei commented on source/distributed/full_distributed_tria.cc:
```diff
@@ -0,0 +1,499 @@
+      const ConstructionData<dim, spacedim> &construction_data)
+    {
+      // check if there are locally relevant coarse-grid cells
+      if (construction_data.cells.empty() == true) // no
```
```suggestion
      if (construction_data.cells.empty() == true)
```

tjhei commented on source/distributed/full_distributed_tria.cc:
```diff
@@ -0,0 +1,499 @@
+          GridGenerator::hyper_cube(*this, 0, 1, false);
+
+          // 2) mark cell as artificial
+          for (auto cell = this->begin(); cell != this->end(); cell++)
```
remove the for loop, as you only have one cell anyways. You can do ``auto cell=this->begin();``.

tjhei commented on source/distributed/full_distributed_tria.cc:
```diff
@@ -0,0 +1,499 @@
+    Triangulation<dim, spacedim>::reinit(
+      const ConstructionData<dim, spacedim> &construction_data)
+    {
+      // check if there are locally relevant coarse-grid cells
```
assert that Tria is empty? You also might want to clear the coarse_cell_id* members.

tjhei commented on source/distributed/full_distributed_tria.cc:
```diff
@@ -0,0 +1,499 @@
+          // 3) setup dummy mapping between locally relevant coarse-grid cells
+          //    and global cells
+          this->coarse_cell_id_to_coarse_cell_index_vector.emplace_back(0, 0);
+          this->coarse_cell_index_to_coarse_cell_id_vector.push_back(0);
```
can we use something like invalid ids here instead?

tjhei commented on source/distributed/full_distributed_tria.cc:
```diff
@@ -0,0 +1,499 @@
+
+          // 3) setup dummy mapping between locally relevant coarse-grid cells
+          //    and global cells
+          this->coarse_cell_id_to_coarse_cell_index_vector.emplace_back(0, 0);
```
why not leave this empty? There is no coarse cell on this processor...

tjhei commented on source/distributed/full_distributed_tria.cc:
```diff
@@ -0,0 +1,499 @@
+          this->coarse_cell_index_to_coarse_cell_id_vector =
+            construction_data.coarse_cell_index_to_coarse_cell_id;
+
+
```
remove some of this whitespace.

tjhei commented on source/distributed/full_distributed_tria.cc:
```diff
@@ -0,0 +1,499 @@
+          // 3) create coarse grid
+          const SubCellData subcelldata;
+          dealii::parallel::Triangulation<dim, spacedim>::create_triangulation(
+            construction_data.vertices, construction_data.cells, subcelldata);
```
can you assert the sizes of the coarse_cell_* variables here?

tjhei commented on source/distributed/full_distributed_tria.cc:
```diff
@@ -0,0 +1,499 @@
+
+          // 4) create all levels via a sequence of refinements
+          const auto &cell_infos = construction_data.cell_infos;
+          for (unsigned int ref_counter = 0; ref_counter < cell_infos.size();
```
is ref_counter the level? Maybe rename?

tjhei commented on source/distributed/full_distributed_tria.cc:
```diff
@@ -0,0 +1,499 @@
+               ref_counter++)
+            {
+              // a) perform refinement
+              if (ref_counter > 0)
```
This confused me a little. Is the code more complicated if you do the refinement for ref_counter==0 except the last?

tjhei commented on source/distributed/full_distributed_tria.cc:
```diff
@@ -0,0 +1,499 @@
+
+                  // boundary ids
+                  for (auto pair : cell_info->boundary_ids)
+                    cell->face(pair.first)->set_boundary_id(pair.second);
```
please add an assert that ``cell->at_boundary(pair.first)``

tjhei commented on source/distributed/full_distributed_tria.cc:
```diff
@@ -0,0 +1,499 @@
+      AssertThrow(false,
+                  ExcMessage(
+                    "Use the function reinit() to create the triangulation!"));
+      dealii::parallel::Triangulation<dim, spacedim>::create_triangulation(
```
Don't call this function if it is not supported.

tjhei commented on source/distributed/full_distributed_tria.cc:
```diff
@@ -0,0 +1,499 @@
+              for (; cell_info != cell_infos[ref_counter].end(); ++cell_info)
+                {
+                  // find cell that has the correct
+                  while (cell_info->id != cell->id().template to_binary<dim>())
```
What happens if you hand in incorrect data and we don't find the cell. Is this an endless loop? Maybe you can check if we arrive at end()?

peterrum replied:
There is not endless loop. Instead, the cell iterator (reaches its end and) tries to access an invalid cell, which is the time an assert is encountered and an assert is thrown.

tjhei commented on source/distributed/full_distributed_tria.cc:
```diff
@@ -0,0 +1,499 @@
+    void
+    Triangulation<dim, spacedim>::execute_coarsening_and_refinement()
+    {
+      Assert(false, ExcNotImplemented());
```
AssertThrow

tjhei commented on source/distributed/full_distributed_tria.cc:
```diff
@@ -0,0 +1,499 @@
+    bool
+    Triangulation<dim, spacedim>::prepare_coarsening_and_refinement()
+    {
+      return dealii::Triangulation<dim, spacedim>::
```
you say it is not supported, but this looks reasonable...

peterrum replied:
I would like to insert here an assert (it does not make sense that this function is called until we do not support refinement). But this is not possible since it is called by `dealii::Triangulation<dim,spacedim>::execute_coarsening_and_refinement()`. I have added a warning to the code.

peterrum commented:
Thanks @tjhei! I will address one by one your comments.

peterrum commented:
@tjhei I have addressed all (but one of) your comments. Thanks a lot. It is much appreciated! 

There is still the section in `parallel::fullydistributed::Triangulation::reinit()` which handles `ConstructionData`s which does not have any locally relevant cells: 
```cpp
      // check if there are locally relevant coarse-grid cells
      if (construction_data.coarse_cell_vertices.empty() == true)
        {
          // 1) create a dummy hypercube
          GridGenerator::hyper_cube(*this, 0, 1, false);

          // 2) mark cell as artificial
          auto cell = this->begin();
          cell->set_subdomain_id(dealii::numbers::artificial_subdomain_id);
          cell->set_level_subdomain_id(
            dealii::numbers::artificial_subdomain_id);

          // 3) setup dummy mapping between locally relevant coarse-grid cells
          //    and global cells
          this->coarse_cell_id_to_coarse_cell_index_vector.emplace_back(0, 0);
          this->coarse_cell_index_to_coarse_cell_id_vector.push_back(0);
        }
      else{ /* Locally relevant cells are available! */ }
```
With @bangerth, we have agreed on removing this section and to require that `ConstructionData` always has at least on cell (even if it artificial). We would pick any cell for the original serial or distributed triangulation. This way we could guarantee that all processes have at least one cell (as required at many places in deal.II) and the cell has some connection to the original geometry (in contrast to an arbitrary `hyper_cube`).

@bangerth  Sorry for bringing up this issue again! But I am not convinced that this is best approach. Let me explain my reasoning: `ConstructionData` is a data structure which has to/can be filled by the user (or by calling a utility function in `p:f:Util`). It is supposed to contain all locally-relevant cells. But in a single case (that a process has no locally relevant cells), it should also contain an arbitrary cell. I guess from a user perspective this does not make any sense (and should be treated internally e.g. here). 

As an alternative, we could also argue with that a user is not supposed to create a `ConstructionData` on his/her own but should use one of the predefined functions (which handles the corner case). This was not what I intended with the introduction of `ConstructionData`: it is supposed to provide more flexibility...

tjhei commented:
> With @bangerth, we have agreed on removing this section and to require that `ConstructionData` always has at least on cell (even if it artificial). We would pick any cell for the original serial or distributed triangulation. This way we could guarantee that all processes have at least one cell (as required at many places in deal.II) and the cell has some connection to the original geometry (in contrast to an arbitrary `hyper_cube`).

I think both options are okay, but I prefer the current implementation. ConstructionData is visible to the user and having to put in dummy or irrelevant data doesn't make much sense to me. You would have to document that.

tjhei commented:
> . This way we could guarantee that all processes have at least one cell (as required at many places in deal.II)

Is this really needed in many places? Would it be enough to override ``empty()`` to return false?

tjhei commented on source/distributed/full_distributed_tria.cc:
```diff
@@ -0,0 +1,508 @@
+
+namespace parallel
+{
+  namespace fullydistributed
```
the namespace and filename do not match.

peterrum replied:
@bangerth wished to change the file name to `full_distributed_tria.cc`. It used to match the namespace (`fullydistributed_tria.cc`).

tjhei replied:
Sorry for the back and forth. :-) I think he meant fully_distributed_tria.cc.

tjhei commented on source/distributed/full_distributed_tria.cc:
```diff
@@ -0,0 +1,508 @@
+    Triangulation<dim, spacedim>::coarse_cell_index_to_coarse_cell_id(
+      const unsigned int coarse_cell_index) const
+    {
+      return coarse_cell_index_to_coarse_cell_id_vector[coarse_cell_index];
```
add an Assert for a valid coarse cell here. Maybe also check that the return value is not an invalid value (as I suggested setting for the one dummy cell).

peterrum commented:
> Is this really needed in many places?

One place I encountered an assert is when calling `tria->begin()`: 
- https://www.dealii.org/developer/doxygen/deal.II/grid_2tria_8cc_source.html#l11897
- https://www.dealii.org/developer/doxygen/deal.II/grid_2tria_8cc_source.html#l12473
- https://www.dealii.org/developer/doxygen/deal.II/grid_2tria_8cc_source.html#l12427

I guess we could check here `empty()` and return `end()`. (And the same iterator functions...) Once these are modified, there might be (or not) new further problematic instances.

peterrum commented:
@tjhei The following files are ready to be reviewed:
- include/deal.II/distributed/fully_distributed_tria_util.h
- source/distributed/fully_distributed_tria_util.cc
- source/distributed/fully_distributed_tria_util.inst.in

Thanks a lot in advance!

I guess we might be able to drop the `limit_level_difference_at_vertices`-flag. However, the serial triangulation and the p:d:t passed to `parallel::fullydistributed::Utilities::copy_from_triangulation` has to have the same flags as p:f:t. I guess I could store the smoothing flags in `ConstructionData` and call `set_mesh_smoothing()` internal within `reinit()`. What do you think about this?

peterrum commented:
I have started to implement global refinement (and coarsening) on a separate branch https://github.com/peterrum/dealii/tree/pft-refinement-coarsening. I will open a PR once this PR is merged.

In file https://github.com/peterrum/dealii/blob/pft-refinement-coarsening/tests/fullydistributed_grids/global_refinement_coarsening_01.cc#L48-L86 you can see an example where I create manually a `ConstructionData`.

As far as I see it at the moment, adaptive meshes are not a problem (if we could somehow communicate the refinement and coarsening flags - any ideas @marcfehling @bangerth how to do this?) in contrast to re-partitioning (which I will leave for the future)...

kronbichler commented on doc/news/changes/minor/20190902PeterMunch:
```diff
@@ -0,0 +1,6 @@
+New: Add a function, which creates a parallel::fullydistributed::ConstructionData from a
+serial and a distributed triangulation. This data structure can be used to initialize
+parallel::fullydistributed::Triangulation. 
+arbitrary dimensions.
```
```suggestion
```

kronbichler commented on include/deal.II/distributed/fully_distributed_tria.h:
```diff
@@ -0,0 +1,401 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2008 - 2019 by the deal.II authors
```
Is there substantial old code that you re-used in this file? If not, please set the copyright year to "2019". It will later get expanded by some script into e.g. "2019 - 2020" if there is a change in 2020, but it cannot be done backwards.

peterrum replied:
I did not know about this. I thought all the headers are the same.

masterleinad replied:
Then, we would need `1998 - ...` everywhere. :wink:

kronbichler commented on include/deal.II/distributed/fully_distributed_tria.h:
```diff
@@ -0,0 +1,401 @@
+     *
+     * To be able to construct a fully partitioned triangulation that
+     * distributes the coarse grid and gives flexibility regarding partitioning,
+     *  the following ingredients are required:
```
```suggestion
     * the following ingredients are required:
```

kronbichler commented on include/deal.II/distributed/fully_distributed_tria.h:
```diff
@@ -0,0 +1,401 @@
+        construct_multigrid_hierarchy = 0x1
+      };
+
+
```
```suggestion
```

kronbichler commented on include/deal.II/distributed/fully_distributed_tria.h:
```diff
@@ -0,0 +1,401 @@
+       * @note The namespace dealii::fullydistributed::Util contains functions
+       *       to create ConstructionData.
+       *
+       * @note This is the function to be used instead of create_triangulation.
```
```suggestion
       * @note This is the function to be used instead of 
       * Triangulation::create_triangulation() for some of the other triangulations
       * of deal.II.
```

kronbichler commented on include/deal.II/distributed/fully_distributed_tria.h:
```diff
@@ -0,0 +1,401 @@
+      reinit(const ConstructionData<dim, spacedim> &construction_data);
+
+      /**
+       * @note This function is not implemented for this class. Instead, use
```
```suggestion
       * @note This function is not implemented for this class and throws an assertion. Instead, use
```

kronbichler commented on include/deal.II/distributed/fully_distributed_tria.h:
```diff
@@ -0,0 +1,401 @@
+
+      /**
+       * This function determines the neighboring subdomains that are adjacent
+       *  to each vertex.
```
```suggestion
       * to each vertex.
```

kronbichler commented on include/deal.II/distributed/fully_distributed_tria.h:
```diff
@@ -0,0 +1,401 @@
+      bool prepare_coarsening_and_refinement_for_internal_usage;
+
+
+      template <typename>
```
Maybe add something like "Make the respective DoFHandler policy a friend."

kronbichler commented on include/deal.II/distributed/fully_distributed_tria_util.h:
```diff
@@ -0,0 +1,68 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2008 - 2019 by the deal.II authors
```
See above. This is definitely new.

kronbichler commented on source/distributed/fully_distributed_tria.cc:
```diff
@@ -0,0 +1,519 @@
+  hyper_cube(Triangulation<dim, spacedim> &tria,
+             const double                  left,
+             const double                  right,
+             const bool                    colorize);
```
Why don't you include `deal.II/grid/grid_generator.h` here?

bangerth replied:
Still a good question ;-)

kronbichler commented on source/distributed/fully_distributed_tria_util.cc:
```diff
@@ -0,0 +1,517 @@
+    {
+      namespace
+      {
+        template <typename CELL>
```
```suggestion
        template <typename CellIteratorType>
```
(We aim to use uniform names for this kind of templates.)

peterrum replied:
Oh, this was pure laziness...

I have changed it to:
```cpp
        template<int dim, int spacedim>
        void
        set_flag_reverse(TriaIterator<CellAccessor<dim, spacedim>> cell);
```

kronbichler commented on source/distributed/fully_distributed_tria_util.cc:
```diff
@@ -0,0 +1,517 @@
+              my_rank = dealii::Utilities::MPI::this_mpi_process(comm);
+            else
+              AssertThrow(false,
+                          ExcMessage("PDT: y_rank has to equal global rank."));
```
Can you please expand PDT? Also, I think it should be `my_rank`.

kronbichler commented on source/distributed/fully_distributed_tria_util.cc:
```diff
@@ -0,0 +1,517 @@
+                      cell_info.manifold_line_ids[line] =
+                        cell->line(line)->manifold_id();
+
+                  // ... of hexes
```
```suggestion
                  // ... of quads
```

kronbichler commented on source/distributed/fully_distributed_tria_util.cc:
```diff
@@ -0,0 +1,517 @@
+                          cell_info.manifold_line_ids[line] =
+                            cell->line(line)->manifold_id();
+
+                      // ... of hex
```
```suggestion
                      // ... of quads
```

kronbichler commented on source/distributed/fully_distributed_tria_util.cc:
```diff
@@ -0,0 +1,517 @@
+                      // ... of cell
+                      cell_info.manifold_id = cell->manifold_id();
+
+                      // ... of line
```
```suggestion
                      // ... of lines
```

kronbichler commented on source/grid/grid_tools_dof_handlers.cc:
```diff
@@ -2063,8 +2063,8 @@ namespace GridTools
-    Assert(pairs1.size() == pairs2.size(),
-           ExcMessage("Unmatched faces on periodic boundaries"));
+    // Assert(pairs1.size() == pairs2.size(),
+    //       ExcMessage("Unmatched faces on periodic boundaries"));
```
How do we proceed with these? (We maybe said something, but I forgot again.)

peterrum replied:
@kronbichler You have suggested to introduce here an if-statement, checking if we are dealing with a `p:f:t` or not. However, I do not have access here to the triangulation, do I?

kronbichler replied:
One could of course try a `dynamic_cast`, but the question is if we have a more basic way to figure these things out. In any case, we ought really keep the assertions for the existing variants.

peterrum replied:
Done!

kronbichler commented on tests/fullydistributed_grids/copy_distributed_tria_01.cc:
```diff
@@ -0,0 +1,80 @@
+main(int argc, char *argv[])
+{
+  Utilities::MPI::MPI_InitFinalize mpi_initialization(argc, argv, 1);
+  MPILogInitAll                    all;
```
It seems you do not write any output from different processors to the output file, so it is a bit useless to collect data from all processors. Then it might be enough to simply use `mpi_initlog();` and ignore everything except the output of rank 0.

kronbichler replied:
Thinking about it, you might want to print the number of dofs (and globally active cells).

tjhei commented on include/deal.II/distributed/fully_distributed_tria.h:
```diff
@@ -0,0 +1,401 @@
+     *       any form of adaptivity, not even simple global refinements and
+     *       coarsenings.
+     *
+     * @note Currently only simple periodicity conditions are supported.
```
can you elaborate what "simple" means?

peterrum replied:
I changed the comment to the following:
```cpp
     * @note Currently only simple periodicity conditions (i.e. without offsets
     *       and rotation matrices - see also the documentation of 
     *       GridTools::collect_periodic_faces)) are supported.
```

peterrum replied:
@tamiko @masterleinad We should tackle this limitation in a future PR!

tjhei commented on include/deal.II/distributed/fully_distributed_tria.h:
```diff
@@ -0,0 +1,401 @@
+      : public parallel::DistributedTriangulationBase<dim, spacedim>
+    {
+    public:
+      typedef typename dealii::Triangulation<dim, spacedim>::cell_iterator
```
we prefer ``using`` over ``typedef``. Can you change this please?

tjhei commented on include/deal.II/distributed/fully_distributed_tria.h:
```diff
@@ -0,0 +1,401 @@
+       *
+       * @note This is the function to be used instead of create_triangulation.
+       *
+       * @param construction_data
```
@param needs a description behind the argument name. Maybe something like "The data needed for this process." or something.

tjhei commented on source/distributed/fully_distributed_tria.cc:
```diff
@@ -0,0 +1,519 @@
+        }
+
+      // check if map is actually compressed
+      for (auto p : periodic_map)
```
Is this something we should only do in debug mode? If yes, surround by ``#ifdef DEBUG``.

tjhei commented on source/distributed/fully_distributed_tria.cc:
```diff
@@ -0,0 +1,519 @@
+    Triangulation<dim, spacedim>::coarse_cell_index_to_coarse_cell_id(
+      const unsigned int coarse_cell_index) const
+    {
+      const auto coarse_cell_id =
```
can you assert ``coarse_cell_index < number of coarse cells``?

tjhei commented on tests/fullydistributed_grids/copy_distributed_tria_01.with_p4est=true.with_metis=true.mpirun=1.output:
```diff
@@ -0,0 +1 @@
+
```
why do you need metis?

peterrum replied:
You are right! These are historical remainders...

tjhei commented on tests/fullydistributed_grids/copy_distributed_tria_01.with_p4est=true.with_metis=true.mpirun=7.output:
```diff
@@ -0,0 +1,13 @@
+
```
why do you need metis?

peterrum replied:
I just noticed that this had a reason: I was using `GridTools::partition_triangulation` with metis as partitioner. I have replaced this function call with `GridTools::partition_triangulation_zorder`, which neither depends on metis nor on zoltan!

tjhei commented on tests/fullydistributed_grids/copy_distributed_tria_02.with_p4est=true.with_metis=true.mpirun=1.output:
```diff
@@ -0,0 +1 @@
+
```
same here

tjhei commented on tests/fullydistributed_grids/copy_serial_tria_01.cc:
```diff
@@ -0,0 +1,98 @@
+
+  try
+    {
+      // deallog.push("1d");
```
is there a problem in 1d?

peterrum replied:
I am testing on a `hyper_L` (somewhat motivated by `parallel::split::Triangulation`), but I guess I additionally can test on a `hyper_cube`!

tjhei commented on tests/fullydistributed_grids/copy_serial_tria_02.with_metis=true.mpirun=1.output:
```diff
@@ -0,0 +1 @@
+
```
can you output some information about multigrid here? Maybe number of cells per level or something?

tjhei commented on tests/fullydistributed_grids/copy_serial_tria_03.cc:
```diff
@@ -0,0 +1,116 @@
+void
+test(int n_refinements, const int n_subdivisions, MPI_Comm comm)
+{
+  // create pdt
```
update comment?

tjhei commented on tests/fullydistributed_grids/copy_serial_tria_04.cc:
```diff
@@ -0,0 +1,110 @@
+// ---------------------------------------------------------------------
+
+
+// create a tria mesh and copy it
```
please extend this one line comment to say what the differences between these tests are.

tjhei commented on tests/fullydistributed_grids/copy_serial_tria_05.cc:
```diff
@@ -0,0 +1,90 @@
+  FE_Q<dim>       fe(2);
+  DoFHandler<dim> dof_handler(tria_pft);
+  dof_handler.distribute_dofs(fe);
+  dof_handler.distribute_mg_dofs();
```
you are also distributing dofs, can you print owned dofs per level or something?

peterrum commented:
@tjhei @kronbichler Thanks for your feedback: I will integrate the suggestion tomorrow or the day after tomorrow!

marcfehling commented:
> As far as I see it at the moment, adaptive meshes are not a problem (if we could somehow communicate the refinement and coarsening flags - any ideas @marcfehling @bangerth how to do this?) in contrast to re-partitioning (which I will leave for the future)...

I am not familiar with all of your features on the `p::fd::Triangulation`, but I assume you need to transfer these flags to ghost cells so that these are up-to-date, since each processor only owns its local domain. You can achieve this with the [`GridTools::exchange_cell_data_to_ghosts()`](https://www.dealii.org/current/doxygen/deal.II/namespaceGridTools.html#a2b0a664be6f80b7ddea4a09564861087) function. But please let me know if you need something different if I misinterpreted your requirements :)

peterrum commented:
@marcfehling sounds as if that will work! Thanks!

marcfehling commented:
We do something similar here, but with `active_fe_indices`: https://github.com/dealii/dealii/blob/df024328c3af415ff77f0e225668173b374065f3/source/hp/dof_handler.cc#L1035-L1066

peterrum commented:
@tjhei @kronbichler Thanks for the comments. It seems we are, indeed, in a good shape and soon ready to merge (ready to test). The plan for the following days are:
- copy the periodic faces from the source triangulation to the `ConstructionData` and use it during `reinit`,
- update the tests (so that they actually output something and do not rely on metis), and
- fix the code for serial compilation.

Anything I have forgotten?

peterrum commented:
@tjhei @marcfehling We should create an issue and have a vote if we use `p:f:Triangulation` (`p:f:t`) or `p:fd:Triangulation` (`p:fd:t`). I vote for `p:f:t`, as far this is not taken yet (In the case of `parallel::split:Triangulation`, we would have problem ;) ).

peterrum commented:
It seems `p:f:t` compiles without problems with `-D DEAL_II_WITH_MPI="OFF"` :) I am now wondering if we should run the test also in the serial case, which would mean to copy the output files, e.g.: `copy_serial_tria_05.mpirun=1.output` -> `copy_serial_tria_05.output`.

tjhei commented:
> It seems `p:f:t` compiles without problems with `-D DEAL_II_WITH_MPI="OFF"` :)

great!

> I am now wondering if we should run the test also in the serial case, which would mean to copy the output files,

yes, that is a good idea.

peterrum commented:
@tjhei My happiness did not last long: The tests fail because of a couple of assert in `parallel::TriangulationBase`:
```cpp
#ifndef DEAL_II_WITH_MPI
    Assert(false,
           ExcMessage("You compiled deal.II without MPI support, for "
                      "which parallel::TriangulationBase is not available."));
#endif
```

We should revisit these asserts at a later time, since there is no reason why `p:f:t` should not work in serial (it is essentially a serial triangulation with just a few added lines)...

peterrum commented:
I would like to postpone the integration of periodic faces in `ConstructionData` to a future PR. I could simply extract `periodic_face_pairs_level_0` from the source triangulation by introducing a new method:
```cpp
/**
  * Return a vector that stores the given periodic face pairs on level 0.
 */
const std::vector<GridTools::PeriodicFacePair<cell_iterator>> &
get_periodic_face_pairs_level_0() const;

template <int dim, int spacedim>
const std::vector<GridTools::PeriodicFacePair<
  typename Triangulation<dim, spacedim>::cell_iterator>> &
Triangulation<dim, spacedim>::get_periodic_face_pairs_level_0() const
{
  return periodic_face_pairs_level_0;
}
```
and use it during `copy_from_triangulation`. This would probably work in the case that I construct all levels (including level 0, which coincides with the source coarse grid). However, in the case only the construction of the active level is requested, I have to translate somehow `periodic_face_pairs_level_0` to the new *coarse level 0*.

For now, this means that `GridTools::collect_periodic_faces()` and `Triangulation::add_periodicity` has to be called twice (once on the source triangulation and once on `p:f:t` once `reinit` was called.)

peterrum commented:
@kronbichler @tjhei Regarding the tests: I have create a new file, which contains common functions of all `p:f:t`-tests (e.g. printing statistics regarding `Triangulation` and `DoFHandler`):

- tests/fullydistributed_grids/tests.h

The functions are called at the end of the tests (at the moment I am doing it only in one test as example):
```cpp
  // print statistics
  print_statistics(tria_pft, true);
  print_statistics(dof_handler, true);
```
The resulting output can be found here:
- tests/fullydistributed_grids/copy_serial_tria_02.mpirun=1.output
- tests/fullydistributed_grids/copy_serial_tria_02.mpirun=7.output

If this is fine this way, I will call the new functions also in the other tests.

kronbichler commented:
/rebuild

peterrum commented:
I think the new tests did not run, did they?

peterrum commented:
@tjhei I had a look at `parallel::split::Triangulation` (PR #3956 - I call it here short `p:s:t`) again. As far as I see it, `parallel::fullydistributed::Triangulation` supports at least all the features you wanted to introduce in that PR. 

- What was called in the case of `p:s:t` `create_mesh_from_marked_cells` is now split up in `Utilities::copy_from_triangulation()` and `reinit()`.
- `coarse_cell_index_to_global_index` and `global_coarse_index_to_cell_index`.
- Methods missing are `n_global_coarse_cells` and `available_coarse_cells`. Do we need these?
- The `DoFHandler::Policy` has been unified since then for distributed triangulations (such that this PR did not have to touch it).

Are you missing something in my implementation? Just as a remark if we would move the `reinit` outside of `p:f:t`, we practically would end up with a serial triangulation ;)

kronbichler commented:
> I think the new tests did not run, did they?

The CI machines do only run a subset of all 11,600 tests: The debug tests on a serial implementation plus in the MPI case only some multigrid transfer cases. This is to keep the turnover times reasonable. If you want, I can run the whole test suite on our "node29" machine which has a reasonable pass rate (around 6-10 failing tests, some of which due to TBB, none of which related to MPI).

kronbichler commented:
> Methods missing are `n_global_coarse_cells` and `available_coarse_cells`. Do we need these?

I do not think we need those at the application level - the only information I could imagine is either the number of active cells or some multigrid information for the coarse level; in the latter case one will have a `DoFHandler` with the most appropriate information.

peterrum commented:
@kronbichler I will run it myself. Thanks!

kronbichler commented:
MKL_HOME=/lnm/intel/mkl/lib/intel64_lin/
/lnm/bin/cmake \
    -D CMAKE_CXX_COMPILER="mpicxx" \
    -D CMAKE_C_COMPILER="mpicc" \
    -D CMAKE_CXX_FLAGS="-march=skylake-avx512 -pthread -fopenmp" \
    -D DEAL_II_LINKER_FLAGS="-fopenmp" \
    -D CMAKE_C_FLAGS="-march=skylake-avx512" \
    -D DEAL_II_CXX_FLAGS_RELEASE="-O3" \
    -D DEAL_II_WITH_MPI:BOOL="ON" \
    -D DEAL_II_WITH_TRILINOS:BOOL="ON" \
    -D TRILINOS_DIR="/home/kronbichler/sw/trilinos/trilinos-12-12-gcc9" \
    -D DEAL_II_WITH_PETSC:BOOL="ON" \
    -D PETSC_DIR="/home/kronbichler/sw/petsc-3.10.5" \
    -D PETSC_ARCH="arch-linux2-c-opt" \
    -D DEAL_II_WITH_METIS:BOOL="ON" \
    -D MPIEXEC_PREFLAGS="-bind-to none" \
    -D METIS_DIR="/home/kronbichler/sw/parmetis-install" \
    -D DEAL_II_WITH_THREADS:BOOL="ON" \
    -D DEAL_II_WITH_NETCDF:BOOL="OFF" \
    -D DEAL_II_FORCE_BUNDLED_BOOST="ON" \
    -D DEAL_II_UNITY_BUILD="ON" \
    -D DEAL_II_WITH_P4EST="ON" \
    -D P4EST_DIR="/home/kronbichler/sw/p4est/build_ompi-4.0.1" \
    -D DEAL_II_COMPONENT_DOCUMENTATION="OFF" \
    -D DEAL_II_STATIC_EXECUTABLE="OFF" \
    -D LAPACK_LIBRARIES="$MKL_HOME/libmkl_gf_lp64.so;$MKL_HOME/libmkl_sequential.so;$MKL_HOME/libmkl_core.so;/usr/lib64/libpthread.so;/usr/lib64/libm.so" \
    /home/kronbichler/sw/dealii_git

kronbichler commented:
But you will likely skip some TPLs (and include others like METIS).

bangerth commented on include/deal.II/distributed/fully_distributed_tria.h:
```diff
@@ -0,0 +1,406 @@
+       * @note Only used for 3D.
+       */
+      std::array<types::material_id, GeometryInfo<dim>::quads_per_cell>
+        manifold_quad_ids;
```
Shouldn't this be `::faces_per_cell`? It's only used in 3d, so it doesn't make a difference. Or maybe just replace "faces" -> "face quads" in the description above to make the equivalence clear.

peterrum replied:
I would like to keep it like this, since above:
```cpp
      std::array<types::material_id, GeometryInfo<dim>::lines_per_cell>
        manifold_line_ids;
```
`lines_per_cell` is used and this way it kind of symmetrical...

bangerth commented on include/deal.II/distributed/fully_distributed_tria.h:
```diff
@@ -0,0 +1,406 @@
+      /**
+       * Mapping from coarse-grid index to coarse-grid id.
+       */
+      std::vector<types::coarse_cell_id> coarse_cell_index_to_coarse_cell_id;
```
Can you elaborate what the two are? Maybe just be referencing the glossary entry I added a while ago.

bangerth replied:
Also, it's a mapping, but a particular one: It's really just a list that, for each locally owned (relevant?) coarse mesh cell provides the corresponding global coarse-grid id. I think that a sentence like that might be easier to read in the description of the field.

bangerth commented on include/deal.II/distributed/fully_distributed_tria.h:
```diff
@@ -0,0 +1,406 @@
+
+
+    /**
+     * Data used to construct a parallel::fullydistributed::Triangulation.
```
Can you either (i) describe what one would store in an object of this kind, or (ii) reference the function that consumes such objects and that describes what one would store in an object of this kind?

bangerth commented on include/deal.II/distributed/fully_distributed_tria.h:
```diff
@@ -0,0 +1,406 @@
+     *   parallel::distributed::Triangulation). Normally, a process only needs a
+     *   small section of the global triangulation, i.e., a small section of the
+     *   coarse grid such that a partitioning of the coarse grid is indeed
+     *   essential. We call cells needed by a process locally relevant cells.
```
Is this different from "locally owned"? I'm asking because that's a well established term in deal.II. It may be useful to make clear whether the "locally relevent cells" require an overlap between processors. You might want to look to the corresponding glossary entries.

peterrum replied:
below it is described as:
```cpp
     * - a locally relevant coarse-grid triangulation
     *   (vertices, cell definition; including a layer of ghost cells)
```

bangerth commented on include/deal.II/distributed/fully_distributed_tria.h:
```diff
@@ -0,0 +1,406 @@
+     *   processes by simply partitioning a space-filling curve might not lead
+     *   to an optimal result for triangulations that originate from large
+     *   coarse grids: e.g. partitions that belong to the same process might
+     *   be discontinuous, leading to increased communication amount (within a
```
```suggestion
     *   be discontinuous, leading to increased communication (within a
```

bangerth commented on include/deal.II/distributed/fully_distributed_tria.h:
```diff
@@ -0,0 +1,406 @@
+     *   coarse grids: e.g. partitions that belong to the same process might
+     *   be discontinuous, leading to increased communication amount (within a
+     *   node and beyond). Graph-based partitioning algorithms might be a sound
+     *   alternative.
```
```suggestion
     *   alternative to the space filling curve used by parallel::distributed::Triangulation.
```

bangerth commented on include/deal.II/distributed/fully_distributed_tria.h:
```diff
@@ -0,0 +1,406 @@
+     * parallel::fullydistributed::ConstructionData. The user has to fill this
+     * data structure - in a pre-processing step - before actually creating the
+     * triangulation. Predefined functions to create ConstructionData
+     * can be found in the namespace dealii::fullydistributed::Util.
```
Can we rename this to `...::Utilities`? We already have several of these that are left unabbreviated.

bangerth commented on include/deal.II/distributed/fully_distributed_tria.h:
```diff
@@ -0,0 +1,406 @@
+     * @note Currently no modifications of the triangulation is supported after
+     *       it has been created, i.e., this type of mesh does not support
+     *       any form of adaptivity, not even simple global refinements and
+     *       coarsenings.
```
Doesn't this contradict the previous statement about hanging nodes and multigrid?

peterrum replied:
You can create a triangulation with hanging nodes and multigrid levels (with `reinit`). However, once it has been created, it cannot be altered anymore, i.e. you cannot coarsen or refine afterwards...

Obviously, my comment above is misleading. Do you have any recommendations how phrase it differently?

bangerth commented on include/deal.II/distributed/fully_distributed_tria.h:
```diff
@@ -0,0 +1,406 @@
+     *
+     * @note Currently only simple periodicity conditions (i.e. without offsets
+     *       and rotation matrices - see also the documentation of
+     *       GridTools::collect_periodic_faces)) are supported.
```
```suggestion
     *       GridTools::collect_periodic_faces()) are supported.
```

bangerth commented on include/deal.II/distributed/fully_distributed_tria.h:
```diff
@@ -0,0 +1,406 @@
+      /**
+       * Constructor.
+       *
+       * @param mpi_communicator denotes the MPI communicator to be used for the
```
```suggestion
       * @param mpi_communicator The MPI communicator to be used for the
```
(Take a look at how doxygen renders these `@param` statements.

peterrum replied:
This was just copy-and-past from `p:d:t` (see also PR #8725).

bangerth commented on include/deal.II/distributed/fully_distributed_tria.h:
```diff
@@ -0,0 +1,406 @@
+       * @param construction_data The data needed for this process.
+       */
+      void
+      reinit(const ConstructionData<dim, spacedim> &construction_data);
```
Why not just call this function `create_triangulation` as well? That would limit the need for learning new interfaces.

bangerth commented on include/deal.II/distributed/fully_distributed_tria.h:
```diff
@@ -0,0 +1,406 @@
+
+      /**
+       * Override the function to update the number cache so we can fill data
+       * like level_ghost_owners.
```
This is a description for those who know the implementation of these classes, but it doesn't provide any useful information to people who only *use* the class. I'd just omit the description, in which case doxygen will copy the description from the base class.

Separately: Shouldn't this function be `private`?

peterrum replied:
- you are right `update_number_cache` should be private
- the comment (copied from `p:d:t`) describes what happens additionally to the base class: so I would keep it...

bangerth commented on include/deal.II/distributed/fully_distributed_tria.h:
```diff
@@ -0,0 +1,406 @@
+      compute_vertices_with_ghost_neighbors() const override;
+
+      /**
+       * @copydoc dealii::parallel::DistributedTriangulationBase::is_multilevel_hierarchy_constructed
```
you can just omit everything between `/** ... */` and doxygen will copy what's in the base class.

bangerth commented on include/deal.II/distributed/fully_distributed_tria.h:
```diff
@@ -0,0 +1,406 @@
+      is_multilevel_hierarchy_constructed() const override;
+
+      /**
+       * @copydoc dealii::coarse_cell_id_to_coarse_cell_index
```
same here and below

bangerth commented on include/deal.II/distributed/fully_distributed_tria.h:
```diff
@@ -0,0 +1,406 @@
+        coarse_cell_id_to_coarse_cell_index_vector;
+
+      /**
+       * List of coarse-cell ids for each coarse cell (stored at cell->index()).
```
This sounds like there may be multiple coarse-cell ids for each of the coarse cells. Can you rephrase?

bangerth commented on include/deal.II/distributed/fully_distributed_tria.h:
```diff
@@ -0,0 +1,406 @@
+       * Boolean indicating that the function create_triangulation() was called
+       * for internal usage.
+       */
+      bool create_triangulation_for_internal_usage;
```
Here and the following variable: This sounds like a verb "create this and that" when you really mean "create_triangulation is currently called for internal usage". Can you rename to make clear this distinction (i.e., that you mean a "status", not a verb)? Maybe "currently_processing_create_triangulation" and similar below?

bangerth commented on include/deal.II/distributed/fully_distributed_tria.h:
```diff
@@ -0,0 +1,406 @@
+       */
+      template <typename>
+      friend class dealii::internal::DoFHandlerImplementation::Policy::
+        ParallelDistributed;
```
You probably don't actually need this. I removed it from `p::d::Triangulation` earlier today.

bangerth commented on include/deal.II/distributed/fully_distributed_tria_util.h:
```diff
@@ -0,0 +1,68 @@
+    {
+      /**
+       * Construct parallel::fullydistributed::ConstructionData from a given
+       * partitioned triangulation tria and a specified process.
```
```suggestion
       * partitioned triangulation `tria` and a specified process.
```

bangerth commented on include/deal.II/distributed/fully_distributed_tria_util.h:
```diff
@@ -0,0 +1,68 @@
+       * The input triangulation can be either
+       * a serial triangulation of type dealii::Triangulation which has been
+       * colored (subdomain_id and/or level_subdomain_id has been set) or a
+       * distributed triangulation of type dealii::distributed::Triangulation.
```
```suggestion
       * distributed triangulation of type dealii::distributed::Triangulation
       * where the partitioning is then taken from the way ....[don't actually know, please describe]...
```

bangerth commented on include/deal.II/distributed/fully_distributed_tria_util.h:
```diff
@@ -0,0 +1,68 @@
+       *
+       * @param tria Partitioned input triangulation.
+       * @param tria_pft A parallel::fullydistributed::Triangulation to be
+       *                 passed to extract the settings.
```
If this is all, why not just receive a `Settings` object?

bangerth commented on include/deal.II/distributed/fully_distributed_tria_util.h:
```diff
@@ -0,0 +1,68 @@
+       *         parallel::fullydistributed::Triangulation.
+       *
+       * @note Multilevel hierarchies are supported if it is enabled in
+       *       parallel::fullydistributed::Triangulation.
```
```suggestion
       *       the settings object passed as second argument.
```

bangerth commented on include/deal.II/distributed/fully_distributed_tria_util.h:
```diff
@@ -0,0 +1,68 @@
+       * @note Multilevel hierarchies are supported if it is enabled in
+       *       parallel::fullydistributed::Triangulation.
+       *
+       * @note Hanging nodes are supported. However, to be able to use this
```
You mean hanging nodes in the *input* triangulation?

bangerth commented on include/deal.II/distributed/fully_distributed_tria_util.h:
```diff
@@ -0,0 +1,68 @@
+       *
+       * @note Hanging nodes are supported. However, to be able to use this
+       *       feature, the user has to enable multilevel hierarchy support in
+       *       parallel::fullydistributed::Triangulation..
```
```suggestion
       *       parallel::fullydistributed::Triangulation.
```

bangerth commented on include/deal.II/distributed/fully_distributed_tria_util.h:
```diff
@@ -0,0 +1,68 @@
+       */
+      template <int dim, int spacedim = dim>
+      ConstructionData<dim, spacedim>
+      copy_from_triangulation(
```
Maybe call it `create_construction_data_from_triangulation`? A function called "copy" will generally have an input and output argument, but this one only has input arguments.

bangerth commented on source/dofs/dof_handler.cc:
```diff
@@ -1293,8 +1293,8 @@ DoFHandler<dim, spacedim>::distribute_mg_dofs()
-  if (dynamic_cast<const parallel::distributed::Triangulation<dim, spacedim> *>(
-        &*tria) == nullptr)
+  if (dynamic_cast<const parallel::DistributedTriangulationBase<dim, spacedim>
+                     *>(&*tria) == nullptr)
```
If you want, put this into a PR of its own. It's correct even without the rest of your patch.

peterrum replied:
Done! See PR #8724.

bangerth commented on source/distributed/fully_distributed_tria.cc:
```diff
@@ -0,0 +1,525 @@
+      template <int dim>
+      bool
+      is_parent(typename CellId::binary_type parent,
+                typename CellId::binary_type child)
```
This is a function that should actually be a member of `CellId`. To make it clear who is a parent of whom, I'd call the function `CellId::is_parent_of()` or `CellId::is_child_of()`.

bangerth replied:
You should probably also make that as a separate PR.

bangerth commented on source/distributed/fully_distributed_tria.cc:
```diff
@@ -0,0 +1,525 @@
+    namespace internal
+    {
+      /**
+       * Check if `parent` is the parent of `child`.
```
Only direct parent, or also grandparent etc?

bangerth commented on source/distributed/fully_distributed_tria.cc:
```diff
@@ -0,0 +1,525 @@
+      {
+        if (child[0] != parent[0])
+          return false; // has same coarse cell
+        if ((child[1] >> 2) != ((parent[1] >> 2) + 1))
```
can you explain what the shift operator does?

I general, the remainder of the function does a lot of bit fiddling that will be incomprehensible to most everyone. It could use a few high-level comments that explain the overall algorithmic flow.

bangerth commented on source/distributed/fully_distributed_tria.cc:
```diff
@@ -0,0 +1,525 @@
+          return false; // level has one difference
+
+
+        const unsigned int n_child_indices = parent[1] >> 2;
```
```suggestion
        const unsigned int n_child_indices = (parent[1] >> 2);
```

bangerth commented on source/distributed/fully_distributed_tria.cc:
```diff
@@ -0,0 +1,525 @@
+      this->coarse_cell_index_to_coarse_cell_id_vector.clear();
+
+      // check if there are locally relevant coarse-grid cells
+      if (construction_data.coarse_cell_vertices.empty() == true)
```
It's a bit hard to read because you ask whether a negative question ("empty") is true :-) I'd say that's one occasion where we could omit the explicit `== true`.

Separately, the comment above states the opposite of what the conditional asks ;-)

bangerth commented on source/distributed/fully_distributed_tria.cc:
```diff
@@ -0,0 +1,525 @@
+          cell->set_level_subdomain_id(
+            dealii::numbers::artificial_subdomain_id);
+
+          // 3) setup dummy mapping between locally relevant coarse-grid cells
```
```suggestion
          // 3) set up dummy mapping between locally relevant coarse-grid cells
```

bangerth commented on source/distributed/fully_distributed_tria.cc:
```diff
@@ -0,0 +1,525 @@
+          //    and global cells
+          this->coarse_cell_id_to_coarse_cell_index_vector.emplace_back(
+            numbers::invalid_coarse_cell_id, 0);
+          this->coarse_cell_index_to_coarse_cell_id_vector.push_back(
```
Since you use `emplace_back` above, just use it here as well (even though it doesn't make a difference)

bangerth commented on source/distributed/fully_distributed_tria.cc:
```diff
@@ -0,0 +1,525 @@
+          this->coarse_cell_index_to_coarse_cell_id_vector =
+            construction_data.coarse_cell_index_to_coarse_cell_id;
+
+          // 2) setup `coarse-cell id to coarse-cell index`-mapping
```
```suggestion
          // 2) set up `coarse-cell id to coarse-cell index`-mapping
```

bangerth commented on source/distributed/fully_distributed_tria.cc:
```diff
@@ -0,0 +1,525 @@
+            coarse_cell_id_to_coarse_cell_index_vector;
+          for (unsigned int i = 0;
+               i < construction_data.coarse_cell_index_to_coarse_cell_id.size();
+               i++)
```
```suggestion
               ++i)
```

bangerth commented on source/distributed/fully_distributed_tria.cc:
```diff
@@ -0,0 +1,525 @@
+          dealii::parallel::Triangulation<dim, spacedim>::create_triangulation(
+            construction_data.coarse_cell_vertices,
+            construction_data.coarse_cells,
+            subcelldata);
```
You don't need the local variable:
```suggestion
            SubCellData());
```

bangerth commented on source/distributed/fully_distributed_tria.cc:
```diff
@@ -0,0 +1,525 @@
+
+          Assert(this->n_cells() ==
+                   this->coarse_cell_id_to_coarse_cell_index_vector.size(),
+                 ExcMessage("Sizes do not match!"));
```
I'd just call it `ExcInternalError()`. Same below -- there is nothing the user did wrong if this assertion fails.

bangerth commented on source/distributed/fully_distributed_tria.cc:
```diff
@@ -0,0 +1,525 @@
+
+          // 4) create all levels via a sequence of refinements
+          const auto &cell_infos = construction_data.cell_infos;
+          for (unsigned int level = 0; level < cell_infos.size(); level++)
```
```suggestion
          for (unsigned int level = 0; level < cell_infos.size(); ++level)
```

bangerth commented on include/deal.II/distributed/fully_distributed_tria.h:
```diff
@@ -0,0 +1,406 @@
+      /**
+       * CellInfo for each locally relevant cell on each level.
+       */
+      std::vector<std::vector<CellInfo<dim>>> cell_infos;
```
I missed at first reading that this is a vector of vectors. You'll have to explain what the outer and inner level mean.

I did read up on this in the implementation, and I have to admit that I don't like it particularly because you have to repeat some much information on the refined levels that can be computed from the coarser parent cells. As I'll point out below, you also have a quadratic loop. I'd much rather you find a different way to describe *how* the coarse mesh should be refined, rather than explicitly listing the details of the cells you want.

bangerth commented on source/distributed/fully_distributed_tria.cc:
```diff
@@ -0,0 +1,525 @@
+    Triangulation<dim, spacedim>::coarse_cell_id_to_coarse_cell_index(
+      const types::coarse_cell_id coarse_cell_id) const
+    {
+      auto coarse_cell_index =
```
```suggestion
      const auto coarse_cell_index =
```

bangerth commented on source/distributed/fully_distributed_tria.cc:
```diff
@@ -0,0 +1,525 @@
+    {
+      Assert(
+        coarse_cell_index < coarse_cell_index_to_coarse_cell_id_vector.size(),
+        ExcMessage("You are trying to access a cell which does not exist!"));
```
Just use ExcIndex.

peterrum commented on include/deal.II/distributed/fully_distributed_tria.h:
```diff
@@ -0,0 +1,406 @@
+     * ConstructionData and used during construction of a
+     * parallel::fullydistributed::Triangulation. This struct stores
+     * the cell id, the subdomain_id and the level_subdomain_id as well as
+     * information related to manifold_id and boundary_id.
```
The `ConstructionData` is only used (internally) during `reinit`. The information is stored in `dealii::Triangulation` and only the mapping between coarse-grid id and coarse-cell id and vice versa are stored explicitly. 

I would keep `ConstructionData` as it is for now. Once we proceed to extend `p:f:t` to adaptive meshes with re-partitioning (and have to construct `ConstructionData` again and again), we can think about ways to compress the information in it.

bangerth replied:
I didn't want to suggest you change `ConstructionData` -- this seems entirely fine with me. I read the documentation as suggesting that this format is also used for permanent storage in the Triangulation class, but may have misread things then.

peterrum replied:
The documentation states:
```cpp
    /**
     * Data used to construct a fully distributed triangulation in
     * parallel::fullydistributed::Triangulation::create_triangulation().
     *
     * @author Peter Munch, 2019
     */
```
So we should be fine...

peterrum commented on include/deal.II/distributed/fully_distributed_tria.h:
```diff
@@ -0,0 +1,406 @@
+       * List of face number and boundary id of all non-internal faces of the
+       * cell.
+       */
+      std::vector<std::pair<unsigned int, types::boundary_id>> boundary_ids;
```
This is how it used to be (more or less). However, @tjhei preferred that we explicitly only store ids at actual boundaries.

peterrum replied:
Me neither... In the usual case the vector will be empty...

bangerth replied:
Let's not worry about it then.

peterrum commented on source/distributed/fully_distributed_tria.cc:
```diff
@@ -0,0 +1,525 @@
+      parallel::Triangulation<dim, spacedim>::update_number_cache();
+
+      if (settings & construct_multigrid_hierarchy)
+        parallel::Triangulation<dim, spacedim>::fill_level_ghost_owners();
```
This is just like in `p:d:t`...

bangerth replied:
Ah, but no longer :-) See #8732.

peterrum replied:
Have changed it! I have the feeling `p:f:t` is a good opportunity to revisit `p:d:t`. This is not the first thing, where I used `p:d:t` as template ;)

bangerth replied:
Indeed. That's why I don't mind patch review -- it teaches me about existing code as well, as I frequently look up something somewhere else. Many of my patches are a result of looking up related information while doing patch review.

peterrum commented on source/distributed/fully_distributed_tria.cc:
```diff
@@ -0,0 +1,525 @@
+            }
+
+          // 4a) set all cells artificial
+          for (auto cell = this->begin(); cell != this->end(); cell++)
```
I have updated the comment. I first clear all (level-)subdomain ids and set them in a second loop.

peterrum commented on source/distributed/fully_distributed_tria.cc:
```diff
@@ -0,0 +1,525 @@
+    void
+    Triangulation<dim, spacedim>::execute_coarsening_and_refinement()
+    {
+      AssertThrow(false, ExcNotImplemented());
```
@bangerth I have changed this. But, I am confused why we do not use `AssertThrow` but `Assert` here. This is not really a check, but a notification that the user is definitely doing something wrong...

peterrum commented on source/distributed/fully_distributed_tria.cc:
```diff
@@ -0,0 +1,525 @@
+
+    template <int dim, int spacedim>
+    std::map<unsigned int, std::set<dealii::types::subdomain_id>>
+    Triangulation<dim, spacedim>::compute_vertices_with_ghost_neighbors() const
```
Inside of `parallel::TriangulationBase` we can find an implementation and a **comment** regarding the correctness of this function:
```cpp
   template <int dim, int spacedim>
   std::map<unsigned int, std::set<::types::subdomain_id>>
   TriangulationBase<dim, spacedim>::compute_vertices_with_ghost_neighbors()
     const
   {
     // TODO: we are not treating periodic neighbors correctly here. If we do
     // we can remove the overriding implementation for p::d::Triangulation
     // that is currently using a p4est callback to get correct ghost neighbors
     // over periodic faces.
```
I would not mind if we/I fix this function (by taking some ideas from my implementation) and then remove the implementation from `p:f:t`.

However, it would be nice if someone could help me and point out what the issue with this function is and if it was already solved by my function. Which test could help me, if this function works properly or not? I guess either @tjhei has implemented this function or knows the right person...

I'll clean up my code so that it will be clear what it does (in addition).

ping @bangerth @kronbichler

bangerth replied:
I see. Let's leave this for a follow-up then. Can you open a github issue that describes the problem?

peterrum replied:
I have added comments and have restructured the code. See also the issue #8734.

peterrum commented:
@bangerth Thank you for your very thorough review! I have tackled most of the issues. There are still some to be tackled (indicated by the lack of thumb up)! I'll let you know once I have finished!

bangerth commented on source/distributed/fully_distributed_tria.cc:
```diff
@@ -0,0 +1,534 @@
+    void
+    Triangulation<dim, spacedim>::execute_coarsening_and_refinement()
+    {
+      Assert(false, ExcNotImplemented());
```
The rule is that things that are programming errors only need to be checked in debug mode (`Assert`). This is one of those cases. On the other hand, things that depend on input files should always be checked (`AssertThrow`).

bangerth commented on source/distributed/fully_distributed_tria_util.cc:
```diff
@@ -0,0 +1,520 @@
+      {
+        template <int dim, int spacedim>
+        void
+        set_flag_reverse(TriaIterator<CellAccessor<dim, spacedim>> cell)
```
I can't figure out from the function name what the function does, and it doesn't have documentation :-)

```suggestion
        set_flag_reverse(const TriaIterator<CellAccessor<dim, spacedim>> &cell)
```

bangerth commented on source/distributed/fully_distributed_tria_util.cc:
```diff
@@ -0,0 +1,520 @@
+              temp = temp * GeometryInfo<dim>::max_children_per_cell + i;
+            }
+
+          return temp;
```
There is not a single comment in this function -- it could use some :-)

peterrum commented:
@bangerth  Only a few issues are left to be tackled:
- rebase
- complete the tests
- have some kind of settings for `create_construction_data_from_triangulation` 
- create a PR for `CellId::is_parent_of()` or `CellId::is_child_of()`
- presort the `CellData`
- document `convert_binary_to_gid`

Hopefully, I have all of these changes today!

peterrum commented:
@kronbichler While I was updating the test, I bumped into another *annoying* assert in `dealii/source/grid/tria.cc`:
```cpp
AssertThrow(((spacedim == 1) && (boundary_nodes == 2)) ||
              (spacedim > 1),
            ExcMessage("The Triangulation has too many end points")); 
```
Any idea how to circumvent this for `p:f:t` in 1D? Can I cast in a serial triangulation to `p:f:t`? (Sounds really a bad idea...)

bangerth commented:
I think you can just remove the assertion. Adding the cast is ok as well and would document when this kind of situation can legitimately happen -- so that may be the better choice.

bangerth commented on source/distributed/fully_distributed_tria_util.cc:
```suggestion
            // 2) process all local and ghost cells: set up needed data
```

bangerth commented on source/distributed/fully_distributed_tria_util.cc:
I think that `cell_infos` is empty here, right? In that case, you could probably write this more easily as
```
  cell_infos.resize(1);
```

bangerth replied:
Or, if you like the `push_back()` better, via `push_back({})`

peterrum replied:
I am going with the `resize`-approach.

bangerth commented on source/distributed/fully_distributed_tria_util.cc:
same here with the alias -- just use the full reference to the variable. (I also don't know what "part" would actually mean ;-)

peterrum replied:
Regarding the name `part`: the terminology used to be different (and also the concept). As you will see in a next PR, there will be master process which creates `ConstructionData`s for some partitions  and send them to the slave-processes.

bangerth commented on source/distributed/fully_distributed_tria_util.cc:
Should you be using `dim` in these lines instead of `spacedim`? I think that everything you do in this function should actually be independent of `spacedim`. Same in the next block for the quads.

bangerth commented on source/distributed/fully_distributed_tria_util.cc:
```suggestion
                  part.emplace_back(cell_info);
```

bangerth commented on source/distributed/fully_distributed_tria_util.cc:
Since you're using the user flags, I think you should save/restore the user flags. There are many places in the library where we do this to ensure that users don't see what we're doing with these flags.

bangerth commented on source/distributed/fully_distributed_tria_util.cc:
```suggestion
DEAL_II_NAMESPACE_CLOSE

```

bangerth commented on source/grid/grid_tools_dof_handlers.cc:
```diff
@@ -2063,8 +2064,24 @@ namespace GridTools
+      constexpr int dim      = CellIterator::AccessorType::dimension;
+      constexpr int spacedim = CellIterator::AccessorType::space_dimension;
+      // for parallel::fullydistributed::Triangulation there might be unmatched
+      // faces on periodic boundaries
```
Is this a restriction you intend to lift eventually? This will prevent some algorithms from working.

peterrum replied:
This is not a restriction. This might happen if a ghost coarse-grid cell has a periodic face and the periodic neighbor is not locally relevant.

bangerth replied:
Hm, but *every* ghost neighbor is, by definition, a ghost cell and consequently locally relevant. No?

peterrum replied:
No. 1D example with 8 cells partitioned into 2 domains and periodicity (at the top: ranks of the cells; at the bottom: labeling with ghost (g) and local (l)):
```
0 1 1 1 1 0 0 0
g l l l l g
```
The left ghost cell has periodic face on the left but its left neighbor (i.e. the most right cell) is not locally relevant.

bangerth replied:
Ah, I see. Do you think you can put a summary of this (in particular, that it only works on the coarse mesh) into the comment here (and if there are copies of this comment, there as well)?

bangerth commented on source/grid/grid_tools_dof_handlers.cc:
```diff
@@ -2100,9 +2117,22 @@ namespace GridTools
-    AssertThrow(n_matches == pairs1.size() && pairs2.size() == 0,
-                ExcMessage("Unmatched faces on periodic boundaries"));
+    // Assure that all faces are matched if not
+    // parallel::fullydistributed::Triangulation is used
```
same here

bangerth commented on source/grid/grid_tools_dof_handlers.cc:
```diff
@@ -2237,8 +2267,16 @@ namespace GridTools
-    Assert(pairs1.size() == pairs2.size(),
-           ExcMessage("Unmatched faces on periodic boundaries"));
+    if (!(((pairs1.size() > 0) &&
+           (dynamic_cast<
```
and here, though you didn't copy the comment from above.

bangerth commented on source/grid/tria.cc:
```diff
@@ -1859,9 +1859,10 @@ namespace internal
-                        (spacedim > 1),
-                      ExcMessage("The Triangulation has too many end points"));
+          // AssertThrow(((spacedim == 1) && (boundary_nodes == 2)) ||
+          //              (spacedim > 1),
```
see the comment elsewhere.

bangerth commented on tests/fullydistributed_grids/CMakeLists.txt:
```diff
@@ -0,0 +1,4 @@
+CMAKE_MINIMUM_REQUIRED(VERSION 2.8.12)
+INCLUDE(../setup_testsubproject.cmake)
+PROJECT(testsuite CXX)
+DEAL_II_PICKUP_TESTS()
```
```suggestion
DEAL_II_PICKUP_TESTS()

```

peterrum commented on source/distributed/fully_distributed_tria_util.cc:
```diff
@@ -0,0 +1,523 @@
+      ConstructionData<dim, spacedim>
+      create_construction_data_from_triangulation(
+        const dealii::Triangulation<dim, spacedim> &tria,
+        const Triangulation<dim, spacedim> &        tria_pft,
```
I agree. It is on my todo-list.

peterrum commented on source/distributed/fully_distributed_tria_util.cc:
```diff
@@ -0,0 +1,523 @@
+        ConstructionData<dim, spacedim> construction_data;
+
+        auto &cells    = construction_data.coarse_cells;
+        auto &vertices = construction_data.coarse_cell_vertices;
```
I do not like this but if that is the way it done...

peterrum commented on source/distributed/fully_distributed_tria_util.cc:
```diff
@@ -0,0 +1,523 @@
+        // helper function, which collects all vertices belonging to a cell
+        // (also taking into account periodicity)
+        auto add_vertices_of_cell_to_vertices_owned_by_loclly_owned_cells =
+          [](auto &cell, auto &vertices_owned_by_loclly_owned_cells) mutable {
```
- I used to need the `mutable` her since the lambda used to have a side-effect.
- It is really annoying that in C++11 lambda and auto do not work together!

bangerth commented:
What's missing by now? Other than the one comment thing, are there other issues that need to be addressed, or can we move forward with it?

kronbichler commented:
> Other than the one comment thing, are there other issues that need to be addressed, or can we move forward with it?

I think we can move forward with this. As a matter of fact, I'm probably not the only one who got github's angry unicorn several times on this PR due to the size of comments, so I think we have reached a very good state. At the same time, we have more than half a year until the next release to iron out things we observed on this very powerful new infrastructure.

@bangerth do you want to make the final merge?

peterrum commented:
@bangerth @kronbichler  I have updated the comments regarding `collect_periodic_faces`.

bangerth commented:
Yes, I do. I've had to reload a dozen times before the page even loads...

bangerth commented:
Well done, @peterrum! We look forward to more patches going in this direction!

Can you rebase all of the other pending patches that build on this and leave a note in their comments that they're ready to look at now?

bangerth commented:
Incidentally, at one point you should add an entry to doc/news/changes/major when you're at a point where you think that it's worth announcing the new class!

peterrum commented:
@bangerth Thanks for all the help! I'll rebase the other PRs right away! By the way, there is already major change-log entry ;)
